### PR TITLE
Created simulator backend

### LINKED ISSUE #2113: Create new SimulatorBackend
### Description of the task

Currently the `WifiBackend` is used when we want run our AI on either physical robots or on simulator robots. The issue is that some features (ex Estop) can only be run on one kind of robots. Instead of making `WifiBackend` more complex it is better to make a new backend dedicated to handling simulator robots. 

### Acceptance criteria

- [ ] Have a working `SimulatorBackend` that is able to communicate with standalone simulator and AI
- [ ] Update documentation

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description
Essentially a copy of `wifiBackend` except that the `SimulatorBackend` uses different multicast ipv6 addresses. 
I added the simulator multicast addresses in `software/constants.h` so that they are separate from firmware. 

### Testing Done

tested manually by running full system with wifiBackend and simulatorBackend. AI vs AI also works. 
### Resolved Issues
resolves #2113 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — DISMISSED** [review]
looks great!

File: src/software/backend/simulator_backend.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * Joins the specified multicast group on the vision_output, primitive_output
+     * and robot_status_input. Multicast Channel and Multicast Group are used
+     * interchangeably.
+     *
+     * NOTE: This will terminate the existing connection on the previous channel
+     * if it exists.
+     *
+     * @param channel The channel to join, index of MULTICAST_CHANNELS in
+     * shared/constants.h
```
Comment (jonathanlew) [inline src/software/backend/simulator_backend.h]: nit: change this comment

Comment (amrm3lm) [reply src/software/backend/simulator_backend.h]: changed and updated the one in wifi_backend.h

**Review by jonathanlew — APPROVED** [review]

File: src/software/constants.h [hunk 2]
```diff
@@ -5,3 +5,19 @@
...
 // considered as "gone" and no longer reported.
 static constexpr unsigned int ROBOT_DEBOUNCE_DURATION_MILLISECONDS = 200;
+
+
+static constexpr unsigned int MAX_SIMULATOR_MULTICAST_CHANNELS = 16;
+
+// Networking
+// the IPv6 multicast address, only ff02 is important, the rest is random
+// see https://en.wikipedia.org/wiki/Solicited-node_multicast_address for why ff02 matters
+static const std::string SIMULATOR_MULTICAST_CHANNELS[MAX_SIMULATOR_MULTICAST_CHANNELS] =
```
Comment (akhilveeraghanta) [inline src/software/constants.h:15]: smart! can't use `std::string` in fw lol

**Review by akhilveeraghanta — APPROVED** [review]
lgtm!

File: src/software/constants.h (hunk 2 above)
Comment (amrm3lm) [reply src/software/constants.h:15]: ;)

**Pull request merged**